MediaBrowser.Model/Dlna/StreamBuilder.cs

```diff
@@ -665,251 +587,409 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
                 DeviceProfile = options.Profile
             };
 
-            playlistItem.SubtitleStreamIndex = options.SubtitleStreamIndex ?? GetDefaultSubtitleStreamIndex(item, options.Profile.SubtitleProfiles);
-            var subtitleStream = playlistItem.SubtitleStreamIndex.HasValue ? item.GetMediaStream(MediaStreamType.Subtitle, playlistItem.SubtitleStreamIndex.Value) : null;
+            outputStream.SubtitleStreamIndex = options.SubtitleStreamIndex ?? GetDefaultSubtitleStreamIndex(item, options.Profile.SubtitleProfiles);
+            var subtitleStream = outputStream.SubtitleStreamIndex.HasValue ? item.GetMediaStream(MediaStreamType.Subtitle, outputStream.SubtitleStreamIndex.Value) : null;
 
             var audioStream = item.GetDefaultAudioStream(options.AudioStreamIndex ?? item.DefaultAudioStreamIndex);
             if (audioStream != null)
             {
-                playlistItem.AudioStreamIndex = audioStream.Index;
+                outputStream.AudioStreamIndex = audioStream.Index;
             }
 
             var videoStream = item.VideoStream;
 
-            // TODO: This doesn't account for situations where the device is able to handle the media's bitrate, but the connection isn't fast enough
-            var directPlayEligibilityResult = IsEligibleForDirectPlay(item, GetBitrateForDirectPlayCheck(item, options, true) ?? 0, subtitleStream, options, PlayMethod.DirectPlay);
-            var directStreamEligibilityResult = IsEligibleForDirectPlay(item, options.GetMaxBitrate(false) ?? 0, subtitleStream, options, PlayMethod.DirectStream);
-            bool isEligibleForDirectPlay = options.EnableDirectPlay && (options.ForceDirectPlay || directPlayEligibilityResult.Item1);
-            bool isEligibleForDirectStream = options.EnableDirectStream && (options.ForceDirectStream || directStreamEligibilityResult.Item1);
-
-            _logger.LogInformation(
-                "Profile: {0}, Path: {1}, isEligibleForDirectPlay: {2}, isEligibleForDirectStream: {3}",
-                options.Profile.Name ?? "Unknown Profile",
-                item.Path ?? "Unknown path",
-                isEligibleForDirectPlay,
-                isEligibleForDirectStream);
+            // First, try to DirectPlay.
+            var (canDirectPlayStream, transcodeReasons) = TryBuildVideoStreamForDirectPlay(item, options, videoStream, audioStream, subtitleStream, outputStream);
```

Comment:
I'm not a fan of having `transcodeReasons` be returned by one method and then altered in other methods. Either they should all return `transcodeReasons` or receive it as input (also not a big fan of side effects, but for private methods it's ok)